p2p/conn/connection_test.go

```diff
@@ -116,6 +119,139 @@ func TestMConnectionStatus(t *testing.T) {
 	assert.Zero(status.Channels[0].SendQueueSize)
 }
 
+func TestMConnectionPongTimeoutResultsInError(t *testing.T) {
+	server, client := net.Pipe()
+	defer server.Close()
+	defer client.Close()
+
+	receivedCh := make(chan []byte)
+	errorsCh := make(chan interface{})
+	onReceive := func(chID byte, msgBytes []byte) {
+		receivedCh <- msgBytes
+	}
+	onError := func(r interface{}) {
+		errorsCh <- r
+	}
+	mconn := createMConnectionWithCallbacks(client, onReceive, onError)
+	err := mconn.Start()
+	require.Nil(t, err)
+	defer mconn.Stop()
+
+	serverGotPing := make(chan struct{})
+	go func() {
+		// read ping
+		server.Read(make([]byte, 1))
+		serverGotPing <- struct{}{}
+	}()
+	<-serverGotPing
+
+	pongTimerExpired := mconn.config.PongTimeout + 10*time.Millisecond
+	select {
+	case msgBytes := <-receivedCh:
+		t.Fatalf("Expected error, but got %v", msgBytes)
+	case err := <-errorsCh:
+		assert.NotNil(t, err)
+	case <-time.After(pongTimerExpired):
+		t.Fatalf("Expected to receive error after %v", pongTimerExpired)
+	}
+}
+
+func TestMConnectionMultiplePongsInTheBeginning(t *testing.T) {
+	server, client := net.Pipe()
+	defer server.Close()
+	defer client.Close()
+
+	receivedCh := make(chan []byte)
+	errorsCh := make(chan interface{})
+	onReceive := func(chID byte, msgBytes []byte) {
+		receivedCh <- msgBytes
+	}
+	onError := func(r interface{}) {
+		errorsCh <- r
+	}
+	mconn := createMConnectionWithCallbacks(client, onReceive, onError)
+	err := mconn.Start()
+	require.Nil(t, err)
+	defer mconn.Stop()
+
+	// sending 3 pongs in a row
+	_, err = server.Write([]byte{packetTypePong})
+	require.Nil(t, err)
+	_, err = server.Write([]byte{packetTypePong})
+	require.Nil(t, err)
+	_, err = server.Write([]byte{packetTypePong})
+	require.Nil(t, err)
+
+	serverGotPing := make(chan struct{})
+	go func() {
+		// read ping
+		server.Read(make([]byte, 1))
+		serverGotPing <- struct{}{}
+		// respond with pong
+		_, err = server.Write([]byte{packetTypePong})
+		require.Nil(t, err)
+	}()
+	<-serverGotPing
+
+	pongTimerExpired := mconn.config.PongTimeout + 10*time.Millisecond
```

Comment:
did this test fail with the old code ? 

my understanding was in this scenario with the previous code, the client should have been deadlocked after the multiple pongs

Reply:
sending multiple pings, not pongs, yes!